source/beamline/BioXAS/BioXASXIAFiltersControl.cpp

```diff
@@ -0,0 +1,252 @@
+					thirdFilterActuator_->canStop() &&
+					fourthFilterActuator_->canStop()
+					);
+	}
+
+	return result;
+}
+
+bool BioXASXIAFiltersControl::validValue(double value) const
+{
+	bool result = false;
+
+	if (value == 0.0 ||
+			value == 0.25 ||
+			value == 0.5 ||
+			value == 0.75 ||
+			value == 1.0 ||
+			value == 1.25 ||
+			value == 1.5 ||
+			value == 1.75 ||
+			value == 10.0
+			)
+		result = true;
+
+	return result;
+}
+
+void BioXASXIAFiltersControl::setFirstFilterActuatorControl(CLSBiStateControl *newControl, double filterThickness)
+{
+	if (firstFilterActuator_ != newControl) {
+
+		if (firstFilterActuator_)
+			removeFilterActuatorControl(firstFilterActuator_);
+
+		firstFilterActuator_ = newControl;
+
+		if (firstFilterActuator_)
+			addFilterActuatorControl(firstFilterActuator_, filterThickness);
+
+		emit firstFilterActuatorControlChanged(firstFilterActuator_);
+	}
+}
+
+void BioXASXIAFiltersControl::setSecondFilterActuatorControl(CLSBiStateControl *newControl, double filterThickness)
+{
+	if (secondFilterActuator_ != newControl) {
+
+		if (secondFilterActuator_)
+			removeFilterActuatorControl(secondFilterActuator_);
+
+		secondFilterActuator_ = newControl;
+
+		if (secondFilterActuator_)
+			addFilterActuatorControl(secondFilterActuator_, filterThickness);
+
+		emit secondFilterActuatorControlChanged(secondFilterActuator_);
+	}
+}
+
+void BioXASXIAFiltersControl::setThirdFilterActuatorControl(CLSBiStateControl *newControl, double filterThickness)
+{
+	if (thirdFilterActuator_ != newControl) {
+
+		if (thirdFilterActuator_)
+			removeFilterActuatorControl(thirdFilterActuator_);
+
+		thirdFilterActuator_ = newControl;
+
+		if (thirdFilterActuator_)
+			addFilterActuatorControl(thirdFilterActuator_, filterThickness);
+
+		emit secondFilterActuatorControlChanged(thirdFilterActuator_);
+	}
+}
+
+void BioXASXIAFiltersControl::setFourthFilterActuatorControl(CLSBiStateControl *newControl, double filterThickness)
+{
+	if (fourthFilterActuator_ != newControl) {
+
+		if (fourthFilterActuator_)
+			removeFilterActuatorControl(fourthFilterActuator_);
+
+		fourthFilterActuator_ = newControl;
+
+		if (fourthFilterActuator_)
+			addFilterActuatorControl(fourthFilterActuator_, filterThickness);
+
+		emit secondFilterActuatorControlChanged(fourthFilterActuator_);
+	}
+}
+
+void BioXASXIAFiltersControl::updateConnected()
+{
+	bool isConnected = (
+				firstFilterActuator_ && firstFilterActuator_->isConnected() &&
+				secondFilterActuator_ && secondFilterActuator_->isConnected() &&
+				thirdFilterActuator_ && thirdFilterActuator_->isConnected() &&
+				fourthFilterActuator_ && fourthFilterActuator_->isConnected()
+				);
+
+	setConnected(isConnected);
+}
+
+bool BioXASXIAFiltersControl::filterIsIn(AMControl *control) const
+{
+	bool result = false;
+
+	CLSBiStateControl *actuator = qobject_cast<CLSBiStateControl*>(control);
+
+	if (actuator && validFilterActuator(actuator))
+		result = int(actuator->value()) == Filter::In;
```

Comment:
Although order of operations makes this okay, I would appreciate some brackets around the comparison for clarity.